### PR TITLE
improvement of the magic breakpoints fix #55

### LINKED ISSUE #55: magic breakpoints problem on XP ntldr
[ntldr.zip](https://github.com/bochs-emu/Bochs/files/12146809/ntldr.zip)

On Windows XP SP3: 

Enable magic-break-points with bochs debugger

It breaks every time, you cant continue (a lot of magic breakpoints):

```
(0) Magic breakpoint
(0) Magic breakpoint
(0) Magic breakpoint
(0) Magic breakpoint
(0) Magic breakpoint
(0) Magic breakpoint
```

```
bochsdbg.exe
========================================================================
                        Bochs x86 Emulator 2.7
              Built from SVN snapshot on August  1, 2021
                Timestamp: Sun Aug  1 10:07:00 CEST 2021
========================================================================
00000000000i[      ] reading configuration from .bochsrc
00000000000e[      ] .bochsrc:782: ataX-master/slave CHS set to 0/0/0 - autodetection enabled
00000000000e[      ] .bochsrc:783: ataX-master/slave CHS set to 0/0/0 - autodetection enabled
00000000000i[      ] Stopping on magic break points
00000000000i[      ] installing win32 module as the Bochs GUI
00000000000i[      ] using log file bochsout.txt
Next at t=0
(0) [0x0000fffffff0] f000:fff0 (unk. ctxt): jmpf 0xf000:e05b          ; ea5be000f0
<bochs:1> c
(0) Magic breakpoint
Next at t=4287923
(0) [0x000000020264] 2000:0000000000000264 (unk. ctxt): jmp .+1  (0x00020267)     ; eb01
<bochs:2> c
(0) Magic breakpoint
Next at t=4570603
(0) [0x000000020257] 2000:0000000000000257 (unk. ctxt): nop                       ; 90
<bochs:3> c
(0) Magic breakpoint
Next at t=4571118
(0) [0x000000020257] 2000:0000000000000257 (unk. ctxt): nop                       ; 90
<bochs:4> c
(0) Magic breakpoint
Next at t=4571699
(0) [0x000000020257] 2000:0000000000000257 (unk. ctxt): nop                       ; 90
<bochs:5> c
(0) Magic breakpoint
Next at t=4572761
(0) [0x000000020257] 2000:0000000000000257 (unk. ctxt): nop                       ; 90
<bochs:6> c
(0) Magic breakpoint
Next at t=4573864
(0) [0x000000020257] 2000:0000000000000257 (unk. ctxt): nop                       ; 90
<bochs:7> c
(0) Magic breakpoint
Next at t=4575039
(0) [0x000000020257] 2000:0000000000000257 (unk. ctxt): nop                       ; 90
<bochs:8>
```

### PR DESCRIPTION
**Pull request by therealdreg**
A good start from PR #55

Just implemented support for various types of magic breakpoints, along with the ability to modify them at runtime from within the debugger. So, the Windows XP NTLDR will no longer be a problem.

fix #55

windows xp ntldr have code like this:

```
0000000000020252: (                    ): mov cr0, eax              ; 0f22c0
0000000000020255: (                    ): xchg bx, bx               ; 87db
0000000000020257: (                    ): nop                       ; 90
```

And this code is called a lot of times!

 ring3 code CANT execute OUT instruction (0x8AE0)

So the ring3-dev cant use magic breakpoints on Bochs debugger (on first instance)

with this PR, the user can select what register should breaks:

```
#=======================================================================
# MAGIC_BREAK:
# This enables the "magic breakpoint" feature when using the debugger.
# The useless cpu instructions XCHG %REGW, %REGW causes Bochs to enter the
# debugger mode. This might be useful for software development.
#
# You can specify multiple at once:
#
# cx dx bx sp bp si di
#
# Example for breaking on "XCHGW %DI, %DI" or "XCHG %SP, %SP" execution
#   magic_break: enabled=1 di sp
#
# If nothing is specified, the default will be used: XCHGW %BX, %BX 
#   magic_break: enabled=1
#
# Note: Windows XP ntldr can cause problems with XCHGW %BX, %BX
#=======================================================================
# magic_break: enabled=1
```

**This PR is 100% backward compatibility**

Added a new command to change from debugger the mask of registers, example adding **XCHGW %DI, %DI** or **XCHGW %BX, %BX**

```
setmagicbps "di bx"
```

clear command:

```
clrmagicbps "di bx"
```

@stlintel @vruppert do you like it?

### CODE REVIEW

**Comment by therealdreg** [comment]
@stlintel more suggestions?

btw, I prefer using "{ }" like:

```
for (int i = 1; i < 8; i++) {
    if (strstr(str, magic_bp_regs[i]) != NULL) {
      new_mask |= (1 << i);
    }
  }
```

Because I saw a lot of security bugs related with this, but I can change if you want

File: bochs/bx_debug/dbg_main.cc
Comment (stlintel) [inline bochs/bx_debug/dbg_main.cc:69]: last suggestion - drop it near to use, meaning right before function bx_dbg_print_magic_bp_mask_from_str

**Review by stlintel — CHANGES_REQUESTED** [review]

File: bochs/bx_debug/dbg_main.cc
Comment (therealdreg) [reply bochs/bx_debug/dbg_main.cc]: Done. If you need any more changes, just let me know. I'm still learning about the Bochs project and the team, but I want to contribute and be useful.

By the way, don't forget to generate the new flex & bison files and upload them to the repo

**Review by stlintel — APPROVED** [review]

**Comment by therealdreg** [comment]
Thank you @stlintel , I'll try not to make the same mistakes in the next PR, and be more respectful of your time and patience!

**Comment by stlintel** [comment]
Hi,

The ideal behavior with github was that once project is approved - you could just merge by itself by pressing merge button in the PR. Or you could set 'Enable auto-merge' in the PR and will be merged automatically. 
I don't know if the merge button is available for you, some people told it is not available ...
How does it look on your side ?

I will merge it on the evening together with lex and yacc files.
For now - let's see if we can make github interface work for you and others !

Thanks,
Stanislav

**Comment by therealdreg** [comment]
> Hi,
> 
> The ideal behavior with github was that once project is approved - you could just merge by itself by pressing merge button in the PR. Or you could set 'Enable auto-merge' in the PR and will be merged automatically. I don't know if the merge button is available for you, some people told it is not available ... How does it look on your side ?
> 
> I will merge it on the evening together with lex and yacc files. For now - let's see if we can make github interface work for you and others !
> 
> Thanks, Stanislav

I think I don't have privileges to do it, can you give me privileges in the repository?

By the way, don't worry, I will never merge anything without your approval, and I'll do everything via PR

File: bochs/bx_debug/parser.y
Comment (stlintel) [inline bochs/bx_debug/parser.y:469]: this will clear all magic breakpoints, is it expected ?
previously you had 'default' on BX

File: bochs/bx_debug/parser.y
Comment (stlintel) [inline bochs/bx_debug/parser.y:482]: this clear all magic breakpoints, match your expectations ?

File: bochs/bx_debug/parser.y
Comment (therealdreg) [reply bochs/bx_debug/parser.y:469]: yes, without args both commands clear all because who needs enable all? I documented this in doc and help, the BX thing for backward comp is **config** file

**Comment by stlintel** [comment]
> I think I don't have privileges to do it, can you give me privileges in the repository?

I an still new to github interface - I wonder how to make it available to everybody.
So once approved PR will be possible to merge without me.

> By the way, don't worry, I will never merge anything without your approval, and I'll do everything via PR

File: bochs/bx_debug/parser.y
Comment (therealdreg) [reply bochs/bx_debug/parser.y:482]: yes, without args both commands clear all because who needs enable all? I documented this in doc and help

**Comment by therealdreg** [comment]
> > I think I don't have privileges to do it, can you give me privileges in the repository?
> 
> I an still new to github interface - I wonder how to make it available to everybody. So once approved PR will be possible to merge without me.
> 
> > By the way, don't worry, I will never merge anything without your approval, and I'll do everything via PR

I have no idea about GIT or GITHUB, but tell me what you want me to try, and we'll figure out if it works

File: bochs/bx_debug/parser.y
Comment (therealdreg) [reply bochs/bx_debug/parser.y:469]: image

File: bochs/bx_debug/parser.y
Comment (therealdreg) [reply bochs/bx_debug/parser.y:482]: image

**Comment by therealdreg** [comment]
@stlintel If what you're thinking is that anyone can merge at any time, that's madness; they could break the repository, and this can cause a lot of conflicts...

Someone with judgment must orchestrate the merge, perform the rebases, or do whatever is necessary to ensure that the history and main branch is OK

For example, imagine the following:

You approve a PR from JOHN's branch that modifies the config.cc file on line 23.
You approve a PR from ANA's branch that also modifies the config.cc file on line 23.

Both have merge privileges... so now what? 

Who resolves the conflict? 

And how??? 

Are you going to let them break the repository / rewriting main history or introduce bugs? Nah

This flow dont makes sense, you have to think about this as if it were a multithreading system xD

**Comment by stlintel** [comment]
github handles the mess.

- I approve both PRs. Let's even assume they both have Auto-Merge enabled.
- One will merge first, don't care which one
- Second will fail to merge to to merge conflict. Author will be required to fix the conflict and get re-approval.
- Once re-approved commit will merge.

I don't know, this is how I used to do at work. There is still a lot to set up in github but for sure it could work super-nicely.

**Comment by therealdreg** [comment]
> is

the problem is the same 

> github handles the mess.
> 
> * I approve both PRs. Let's even assume they both have Auto-Merge enabled.
> * One will merge first, don't care which one
> * Second will fail to merge to to merge conflict. Author will be required to fix the conflict and get re-approval.
> * Once re-approved commit will merge.
> 
> I don't know, this is how I used to do at work. There is still a lot to set up in github but for sure it could work super-nicely.

The problem is the same... 

as I said ... someone with judgment must orchestrate the merge, perform the rebases, or do whatever is necessary to ensure that the history and main branch is OK

think in this another scenario:

You approve a PR from JOHN's branch that modifies the config.cc
You approve a PR from ANA's branch that also modifies the dbg_main.cc

Here we don't have a Git conflict, but we might encounter a bug that only a human could detect. What happens if you've tested multiple PRs that may seem 'isolated' or unrelated, but could cause a bug? Only a human can catch something like this... because you are merging in order etc.

**Comment by therealdreg** [comment]
> is

the problem is the same 

> github handles the mess.
> 
> * I approve both PRs. Let's even assume they both have Auto-Merge enabled.
> * One will merge first, don't care which one
> * Second will fail to merge to to merge conflict. Author will be required to fix the conflict and get re-approval.
> * Once re-approved commit will merge.
> 
> I don't know, this is how I used to do at work. There is still a lot to set up in github but for sure it could work super-nicely.

The problem is the same... 

as I said ... someone with judgment must orchestrate the merge, perform the rebases, or do whatever is necessary to ensure that the history and main branch is OK

think in this another scenario:

You approve a PR from JOHN's branch that modifies the config.cc
You approve a PR from ANA's branch that also modifies the dbg_main.cc

Here we don't have a Git conflict, but we might encounter a bug that only a human could detect. What happens if you've tested multiple PRs that may seem 'isolated' or unrelated, but could cause a bug? Only a human can catch something like this... because you are merging in order etc.

**Comment by stlintel** [comment]
>This is key, for your work its very useful this flow... but Bochs its huge and complex.. the contributors like me only knows a little part.. its not the same context IMO

At work the project is way more complex than Bochs, which includes dozens of modules with more than 2M of lines of code.
And these problems can occur there and even do occur, but they are VERY rare.
In the other hand, by merging every branch personally I obviously create a bottleneck.
I could review branch but also other admins could. Volker and hopefully in future, you too.
And if there is an issue -> it could be fixed or entire PR could be reverted.

>In my opinion, only experienced developers with good judgment and deep understand of the project should be able to merge

This is basically the same. Only experienced developers with good judgment are able to approve.
And they should see all PRs and foresee such conflicts.

**Comment by therealdreg** [comment]
> > This is key, for your work its very useful this flow... but Bochs its huge and complex.. the contributors like me only knows a little part.. its not the same context IMO
> 
> At work the project is way more complex than Bochs, which includes dozens of modules with more than 2M of lines of code. And these problems can occur there and even do occur, but they are VERY rare. In the other hand, by merging every branch personally I obviously create a bottleneck. I could review branch but also other admins could. Volker and hopefully in future, you too. And if there is an issue -> it could be fixed or entire PR could be reverted.
> 
> > In my opinion, only experienced developers with good judgment and deep understand of the project should be able to merge
> 
> This is basically the same. Only experienced developers with good judgment are able to approve. And they should see all PRs and foresee such conflicts.

Oka, so if you need something just ping me!

**Review by stlintel — APPROVED** [review]

**Pull request merged**
